AI #2117: Refactor format of CommitmentProgramEligibilityDetails (#2283)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/commitmentprogrameligibilitydetails.md

# Commitment Program Eligibility Details

> Commitment Program Eligibility Details identifies the [*commitment programs*](#glossary:commitment-program) that could potentially cover [*charges*](#glossary:charge), subject to [*service provider*](#glossary:service-provider) constraints. By distinguishing the pool of spend that was eligible to be covered, Commitment Program Eligibility Details provides the fundamental denominator for calculating precise commitment coverage metrics. This allows FinOps practitioners to accurately size the pool of uncovered spend that could realistically be covered by a future commitment. In this context, *commitment programs* include both [*commitment discounts*](#glossary:commitment-discount) and [*capacity reservations*](#glossary:capacity-reservation), provided the service provider treats them as [*commitments*](#glossary:commitment).

Reviewer: The requirements encompass programs beyond **commitment discounts** and **capacity reservations**, but the intro now only names those two.

Can you please confirm this narrowing from **discount-bearing programs** to **commitment discounts** (publicly available, continuous, no carry-over, etc.) in the intro is intentional?

Author: I didn't mean to make a material change; just wanted to root into the glossary terms, not fictitious examples.  How about:

```suggestion
Commitment Program Eligibility Details identifies the [*commitment programs*](#glossary:commitment-program) that could potentially cover [*charges*](#glossary:charge), subject to [*service provider*](#glossary:service-provider) constraints. By distinguishing the pool of spend that was eligible to be covered, Commitment Program Eligibility Details provides the fundamental denominator for calculating precise commitment coverage metrics. This allows FinOps practitioners to accurately size the pool of uncovered spend that could realistically be covered by a future commitment. In this context, *commitment programs* include discount-bearing programs such as [*commitment discounts*](#glossary:commitment-discount) and [*capacity reservations*](#glossary:capacity-reservation), provided the service provider treats them as [*commitments*](#glossary:commitment).
```

Reviewer 2: We talked about this during TF-1 today (2026-04-28). My initial thought was to bring back discount-bearing language and lean on fictitious commitment programs, but including commitments also gives the space to data generators that the initial scope was meant to cover (but was missing).

Author: Great, thanks both!  Committing.

# Commitment Program Eligibility Details

> Commitment Program Eligibility Details identifies the [*commitment programs*](#glossary:commitment-program) that could potentially cover [*charges*](#glossary:charge), subject to [*service provider*](#glossary:service-provider) constraints. By distinguishing the pool of spend that was eligible to be covered, Commitment Program Eligibility Details provides the fundamental denominator for calculating precise commitment coverage metrics. This allows FinOps practitioners to accurately size the pool of uncovered spend that could realistically be covered by a future commitment. In this context, *commitment programs* include both [*commitment discounts*](#glossary:commitment-discount) and [*capacity reservations*](#glossary:capacity-reservation), provided the service provider treats them as [*commitments*](#glossary:commitment).

Author: @Matt-Cowsert @SanjnaSrivatsaVMware 
When the fictitious commitment programs were globally applied, it left this last sentence reading very strangely.  Referencing fake commitment programs here doesn't make any sense, IMO.

I've cleaned it up.  Give me a thumbs up that this looks good to you, and we can resolve.

Reviewer 3: Looks good to me! Can resolve

## Requirements

### Column Requirements

CommitmentProgramEligibilityDetails MUST adhere to the following requirements:

* CommitmentProgramEligibilityDetails MUST be of type JSON Object (serialized as a String where necessary).
* CommitmentProgramEligibilityDetails MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* CommitmentProgramEligibilityDetails MUST conform to [JsonObjectFormat](#attributes.jsonobjectformat) requirements.

> * CommitmentProgramEligibilityDetails MUST conform to [CommitmentProgramEligibilityDetailsObject](#datasets.costandusage.commitmentprogrameligibilitydetails.commitmentprogrameligibilitydetailsobject) requirements when CommitmentProgramEligibilityDetails is not null.

Reviewer 2: ```suggestion
* CommitmentProgramEligibilityDetails MUST conform to [JsonObjectFormat](#attributes.jsonobjectformat) requirements.
```

The bullet being added at line 14 duplicates the pre-existing bullet at line 20. Peer column files (`allocatedmethoddetails.md`, `contractapplied.md`) place the `MUST conform to [...Object] requirements when [...] is not null` bullet at the end of the Column Requirements list, after the nullability bullets. Removing the new insertion here preserves that placement and matches the peer pattern.

Author: Damn!  Rookie mistake.  Accepting.

CommitmentProgramEligibilityDetailsObject MUST adhere to the following requirements:

* CommitmentProgramEligibilityDetailsObject MUST conform to the [CommitmentProgramEligibilityDetailsObjectSchema](#schemas.datasets.costandusage.commitmentprogrameligibilitydetailsobjectschema) JSON Schema.
* CommitmentProgramEligibilityDetailsObject.CommitmentPrograms[\*].ProgramType MUST correspond to a *commitment program* type supported by the service provider.

> * CommitmentProgramEligibilityDetailsObject.CommitmentPrograms[\*].ProgramType MUST equal [CommitmentDiscountType](#datasets.costandusage.commitmentdiscounttype) for one object in CommitmentProgramEligibilityDetailsObject.CommitmentPrograms when CommitmentDiscountType is not null.

Reviewer: Per our Editorial Style Guidelines (Comparison Terminology), string-to-string comparisons should use "match", not "equal":

> **Identifiers and Strings:** Use "match" when comparing strings, IDs, or names (e.g., "HostProviderName MUST match ServiceProviderName"). Do not use "equal".

```suggestion
* CommitmentProgramEligibilityDetailsObject.CommitmentPrograms[\*].ProgramType MUST match [CommitmentDiscountType](#datasets.costandusage.commitmentdiscounttype) for one object in CommitmentProgramEligibilityDetailsObject.CommitmentPrograms when CommitmentDiscountType is not null.
```

|:-------------|:-------------|:-------------|:------------------------------|
| ProgramType | [String](#attributes.stringhandling) | True     | The specific type of commitment program (e.g., discount or capacity reservation) available for this usage. |

## Object Example

Here is a basic example of the object format.

* For more detailed examples, please see this column's entry in the JSON Object Examples appendix entry [here](#appendix.examples:jsonobject.examples:commitmentprogrameligibilitydetails).
* For the JSON schema, please see [Commitment Program Eligibility Details Object Schema](#schemas.datasets.costandusage.commitmentprogrameligibilitydetailsobjectschema).

``` json
{
  "CommitmentPrograms": [
    { "ProgramType": "FlexibleSpendPlan" },
    { "ProgramType": "ResourceReservation" },
    { "ProgramType": "AdvanceResourceCommitment" },
{
  "CommitmentPrograms": [
    { "ProgramType": "Flexible Spend Plan" },
    { "ProgramType": "Resource Reservation" },
    { "ProgramType": "Advance Resource Commitment" },
    { "ProgramType": "Zonal Resource Commitment" }
  ]

> }

Reviewer 2: ```suggestion
{
  "CommitmentPrograms": [
    { "ProgramType": "Flexible Spend Plan" },
    { "ProgramType": "Resource Reservation" },
    { "ProgramType": "Advance Resource Commitment" },
    { "ProgramType": "Zonal Resource Commitment" }
  ]
}
```

The current state has two unclosed JSON objects stacked on top of each other (the original PascalCase array opened but never closed before the new Title Case block began). Collapsing both partial blocks into one valid object matches the examples file and #2266.

Author: Thanks @Matt-Cowsert -- but I can't commit it, as I'm getting a "Applying suggestions on deleted lines is currently not supported" message.

Can you reissue this comment that doesn't include any deleted lines?  Just put the comment at the very end of this block.

Reviewer 2: GitHub keeps blocking the suggestion because the diff hunk for this area is large (131 lines on the left, 34 on the right) and contains the original ZonalResourceCommitment line, the `]`, the `}`, and the entire JSON Type Definition section as deletions. Any suggestion that spans the broken block inherits those deletions, even if I anchor to a smaller range inside the hunk.

Easiest path is a manual replace. Drop in this block in place of lines 58-70:

```json
{
  "CommitmentPrograms": [
    { "ProgramType": "Flexible Spend Plan" },
    { "ProgramType": "Resource Reservation" },
    { "ProgramType": "Advance Resource Commitment" },
    { "ProgramType": "Zonal Resource Commitment" }
  ]
}
```

The opening ` ```json ` fence at line 57 and the closing ` ``` ` at line 71 stay.

Author: Fixed!  Please review and confirm.

> }

Reviewer 2: Per Title Case direction from #2266 (moved here to avoid a merge collision with this refactor). Matches the spacing used in the Cost and Usage CSV examples in #2266.

```suggestion
{
  "CommitmentPrograms": [
    { "ProgramType": "Flexible Spend Plan" },
    { "ProgramType": "Resource Reservation" },
    { "ProgramType": "Advance Resource Commitment" },
    { "ProgramType": "Zonal Resource Commitment" }
  ]
}
```
